File: C++/Binary-Tree/Lowest Common Ancestor Binary Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
        if ( root == NULL )
            return NULL;
        if ( root == p || root == q)
            return root;
        root->left = lowestCommonAncestor( root->left, p, q);
        root->right = lowestCommonAncestor(root->right,p,q);
        if ( root->left != NULL && root->right != NULL)
            return root;
        else
            return ( root->left ? root->left : root->right);
        
    }
};

File: Fb_ph/Practice/Python/Lowest Common Ancestor Binary Tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None
'''
Approach:- Basic Recursive Approach
TC:- O(n)
SC:- O(n)
'''
class Solution(object):
    def lowestCommonAncestor(self, root, p, q):
        """
        :type root: TreeNode
        :type p: TreeNode
        :type q: TreeNode
        :rtype: TreeNode
        """
        def LCA(root,p,q):
            if root == None:
                return
            
            if root == p or root == q:
                return root
            
            root.left = LCA(root.left,p,q)
            root.right = LCA(root.right,p,q)
            
            '''
            Did not understand the below two conditions
            '''
            if root.left and root.right:
                return root
            else:
                root.left = root.left if root.left else root.right
                return root.left
        
        return LCA(root,p,q)
